web/src/torrent.js

```diff
@@ -375,37 +373,38 @@ export class Torrent extends EventTarget {
     }
   }
 
-  /**
-   * @param state one of Prefs.Filter*
-   * @param tracker tracker name
-   * @param search substring to look for, or null
-   * @param labels array of labels. Empty array matches all.
-   * @return true if it passes the test, false if it fails
-   */
-  test(state, tracker, search, labels) {
-    // filter by state...
-    let pass = this.testState(state);
-
-    // maybe filter by text...
-    if (pass && search) {
-      pass = this.getCollatedName().includes(search.toLowerCase());
-    }
-
-    // maybe filter by labels...
-    if (pass) {
-      // pass if this torrent has any of these labels
-      const torrent_labels = this.getLabels();
-      if (labels.length > 0) {
-        pass = labels.some((label) => torrent_labels.includes(label));
+  test(_filter) {
+    let v = null;
+    const pass = (a, vf, callback) => {
```

Comment:
Just a coding style discussion.

Why enclose `v` in `pass()`, and calculate the value of `v` inside pass? It seems more straightforward to pass `v` into `pass()` directly.

Reply:
It's for callback function, calls only when needed, then start another callback function but it will need to retrieve first callback function's return. I know it looks weird but do you know of another way?

Reply:
Now that you said it, I agree that lazy evaluation is a good idea. Still, how about something like this?

```patch
diff --git a/web/src/torrent.js b/web/src/torrent.js
index 31570a59e..2289c15df 100644
--- a/web/src/torrent.js
+++ b/web/src/torrent.js
@@ -374,35 +374,34 @@ export class Torrent extends EventTarget {
   }
 
   test(_filter) {
-    let v = null;
     const pass = (a, vf, callback) => {
       if (a.length === 0) {
         return true;
       }
-      v = vf();
-      return a.some((t) => t.every((x) => callback(x)));
+      const v = vf();
+      return a.some((t) => t.every((x) => callback(x, v)));
     };
 
     return (
       pass(
         _filter.search,
         () => this.getCollatedName(),
-        (x) => v.includes(x),
+        (x, v) => v.includes(x),
       ) &&
       pass(
         _filter.labels,
         () => this.getLabels(),
-        (x) => v.some((z) => z.includes(x)),
+        (x, v) => v.some((z) => z.includes(x)),
       ) &&
       pass(
         _filter.states,
         () => this.getStatus(),
-        (x) => this.testState(v, x),
+        (x, v) => this.testState(v, x),
       ) &&
       pass(
         _filter.trackers,
         () => this.getCollatedTrackers(),
-        (x) => v.includes(x),
+        (x, v) => v.includes(x),
       )
     );
   }
```

Reply:
Jackpot! Why I'm not thinking of that.